### PR TITLE
Create Controlled Paginators

### LINKED ISSUE #1609: Terra-Paginator needs to sync new selectedPage value
# Issue Description
Currently the Paginator does not re-sync the selectedPage props when that prop is updated. We have a search box which rebinds the information and replaces it in the table.

screen shot 2018-06-15 at 2 33 56 pm

## Issue Type
- [ ] Bug

## Expected Behavior

Terra should implement a `componentWillReceiveProps ` or the new `getDerivedStateFromProps` to sync the selectedPage to the correct value

## Steps to Reproduce
1. Create a function that will change how many items are in the paginator
2. Go to the next page
3. Change the function to return less or more items, essentially filtering them
4. Component is in an incorrect state

## Environment
* Component Version: 1.4.0
* Browser Name and Version: All
* Operating System and version (desktop or mobile): All

### PR DESCRIPTION
**Pull request by JakeLaCombe**
### Summary
Fixes #1609

### CODE REVIEW

File: packages/terra-paginator/src/ControlledPaginator.jsx [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    pageSequence.forEach((val) => {
+      const paginationLinkClassNames = cx([
+        'nav-link',
+        val === selectedPage ? 'is-selected' : null,
+      ]);
+
+      if (val > totalPages) {
+        return;
+      }
+      /* eslint-disable comma-dangle, jsx-a11y/aria-props */
```
Comment (Blackop778) [inline packages/terra-paginator/src/ControlledPaginator.jsx]: If you add a set of parenthesis to surround the JSX it will allow you to put the opening `<a>` tag on it's own line and resolve the issues surrounding the tags needing to be on the same column and the need to have a dangling comma.

File: packages/terra-paginator/package.json [hunk 2]
```diff
@@ -50,6 +50,6 @@
     "props-table": "props-table ./src/Paginator.jsx --out-dir ./docs/props-table",
     "test": "npm run test:jest && npm run test:wdio",
     "test:jest": "jest --config ../../jestconfig.js",
-    "test:wdio": "../../node_modules/.bin/wdio ../../wdio.conf.js"
+    "test:wdio": "../../node_modules/.bin/wdio ../../node_modules/terra-dev-site/config/wdio/wdio.conf.js"
```
Comment (bjankord) [inline packages/terra-paginator/package.json]: Curious, why the change here?

Comment (JakeLaCombe) [reply packages/terra-paginator/package.json]: When I run npm install and try to run tests in this package, terra-paginator has trouble finding the wdio.conf.js file as it no longer exists. After looking through other packages, their wdio command looks like this, so a tool kit change must have impacted this.

**Comment by CoryMcDonald** [comment]
Can you update the master README with the component links? The paginator component isn't there 😄

**Comment by JakeLaCombe** [comment]
We actually don't have a master readme for this component. The default readme correlates to the regular paginator, and since the links to the other components exist on the sidebar navigation, I think it would be redundant to add it there.

**Comment by CoryMcDonald** [comment]
screen shot 2018-06-20 at 2 34 13 pm

I meant this @JakeLaCombe  😅

**Comment by JakeLaCombe** [comment]
Gotcha! Good call!

File: packages/terra-paginator/docs/ControlledProgressiveREADME.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# Terra Controlled Progressive Paginator
+
+Controlled version of the [Progressive Paginator](https://github.com/cerner/terra-core/packages/terra-paginator/docs/ProgressiveREADME.md) for when the pagination state needs to be managed by a parent component.
```
Comment (mjhenkes) [inline packages/terra-paginator/docs/ControlledProgressiveREADME.md]: Maybe call out the differences between the api's here?

**Comment by saedar** [comment]
Minor: Your examples have the `<Code>` button overlapping the Dialog footer.

**Comment by bjankord** [comment]
Functional verification completed. The paginator examples display and function correctly.
